signac/contrib/job.py

```diff
@@ -24,31 +23,170 @@
+        # Job objects can point to the same data on disk. We need to store
+        # these jobs in a shared list here so that shallow copies can point to
+        # the same place and trigger each other to update.
+        self._jobs = list(jobs)
+        super().__init__(
+            filename=filename,
+            write_concern=write_concern,
+            data=data,
+            parent=parent,
+            *args,
+            **kwargs,
+        )
 
-    def load(self):
+    def _load(self):
+        # State points never load from disk automatically. They are either
+        # initialized with provided data (e.g. from the state point cache), or
+        # they load from disk the first time state point data is requested for
+        # a Job opened by id (in which case the state point must first be
+        # validated manually).
         pass
 
-    def save(self):
-        """Reset the state point for all the jobs."""
-        for job in self.jobs:
-            job._reset_sp()
+    def _save(self):
+        # State point modification triggers job migration for all jobs sharing
+        # this state point (shallow copies of a single job).
+        new_id = calc_id(self)
+
+        # All elements of the job list are shallow copies of each other, so any
+        # one of them is representative.
+        job = self._jobs[0]
+        old_id = job._id
+        if old_id == new_id:
+            return
+
+        tmp_statepoint_file = self.filename + "~"
+        should_init = False
+        try:
+            # Move the statepoint to an intermediate location as a backup.
+            os.replace(self.filename, tmp_statepoint_file)
+            try:
+                new_workspace = os.path.join(job._project.workspace(), new_id)
+                os.replace(job.workspace(), new_workspace)
+            except OSError as error:
+                os.replace(tmp_statepoint_file, self.filename)  # rollback
+                if error.errno in (errno.EEXIST, errno.ENOTEMPTY, errno.EACCES):
+                    raise DestinationExistsError(new_id)
+                else:
+                    raise
+            else:
+                should_init = True
+        except OSError as error:
+            # The most likely reason we got here is because the state point
+            # file move failed due to the job not being initialized so the file
+            # doesn't exist, which is OK.
+            if error.errno != errno.ENOENT:
+                raise
+
+        # Update each job instance.
+        for job in self._jobs:
+            job._id = new_id
+            job._initialize_lazy_properties()
+
+        # Since all the jobs are equivalent, just grab the filename from the
+        # last one and init it. Also migrate the lock for multithreaded support.
+        old_lock_id = self._lock_id
+        self._filename = job._statepoint_filename
+        type(self)._locks[self._lock_id] = type(self)._locks.pop(old_lock_id)
```

Comment:
Yes, exactly. Per-collection locks need to be implemented by each backend since each backend has a distinct way of identifying a unique resource. For file-based backends like JSON, the unique identifier is the filename. I decided not to document this here since anyone trying to understand (or modify) this code would have to read the docs for the `_lock_id` property anyway.